src/team_data.rs

```diff
@@ -84,24 +92,68 @@ impl TeamApiClient {
     }
 
     pub async fn zulip_map(&self) -> anyhow::Result<ZulipMapping> {
-        download(&self.client, &self.base_url, "/zulip-map.json")
-            .await
-            .context("team-api: zulip-map.json")
+        self.zulip_mapping.get(&self.client, &self.base_url).await
     }
 
     pub async fn teams(&self) -> anyhow::Result<Teams> {
-        download(&self.client, &self.base_url, "/teams.json")
-            .await
-            .context("team-api: teams.json")
+        self.teams.get(&self.client, &self.base_url).await
     }
 
     pub async fn people(&self) -> anyhow::Result<People> {
-        download(&self.client, &self.base_url, "/people.json")
-            .await
-            .context("team-api: people.json")
+        self.people.get(&self.client, &self.base_url).await
     }
 }
 
+/// How long should downloaded team data items be cached in memory.
+const CACHE_DURATION: Duration = Duration::from_secs(2 * 60);
+
+#[derive(Clone)]
+struct CachedTeamItem<T> {
+    value: Arc<RwLock<CachedValue<T>>>,
+    url_path: String,
+}
+
+impl<T: DeserializeOwned + Clone> CachedTeamItem<T> {
+    fn new(url_path: &str) -> Self {
+        Self {
+            value: Arc::new(RwLock::new(CachedValue::Empty)),
+            url_path: url_path.to_string(),
+        }
+    }
+
+    async fn get(&self, client: &Client, base_url: &str) -> anyhow::Result<T> {
+        let now = Instant::now();
+        {
+            let value = self.value.read().await;
+            if let CachedValue::Present {
+                value,
+                last_download,
+            } = &*value
+            {
+                if *last_download + CACHE_DURATION > now {
+                    return Ok(value.clone());
```

Comment:
Team data can be quite big, it seems a bit wasteful to clone the previous data we have in the cache. I supposed we can easily get a reference out of it.

Reply:
We store the values in a thread-safe RwLock, so that it can be accessed behind `&self`. We would need to extend the duration of the lock if we returned references, and essentially return some `LockGuard` objects. I think that's really not worth it here. Before, the code did network roundtrip + deserialize + build a new instance of the team data structure everytime triagebot accessed something. Now it's just a clone. I think it's fine performance-wise.

Reply:
I shouldn't write review comments when I'm tiered. I wanted to say: "**can't** easily get a reference out of it".

> Now it's just a clone. I think it's fine performance-wise.

Yeah, it's probably fine.